returnn/datasets/postprocessing.py

```diff
@@ -0,0 +1,174 @@
+"""
+Provides :class:`PostprocessingDataset`.
+"""
+
+from __future__ import annotations
+
+from numpy import ndarray
+from typing import Any, Callable, Dict, Iterator, List, Optional, Tuple
+
+from returnn.datasets.basic import DatasetSeq
+from returnn.tensor import Tensor, TensorDict
+from .basic import Dataset, init_dataset
+from .cached2 import CachedDataset2
+from .util.strings import str_to_numpy_array
+
+__all__ = ["PostprocessingDataset"]
+
+
+class PostprocessingDataset(CachedDataset2):
+    """
+    A dataset that allows for generic post-processing of data from another dataset
+    using a function on the segment level and on the level of multiply segments via
+    an iterator.
+
+    This allows integrating various data augmentation techniques like e.g. Mixup, or
+    speed pertubation into the data loading pipeline.
+
+    The integration into the data loading pipeline makes it easy to distribute the
+    data processing work across multiple CPU cores using `MultiProcDataset` and in
+    turn frees the GPU from data preprocessing tasks.
+
+    Both functions can be specified at the same time, and their results will be
+    combined by first applying the segment-level function and then the
+    multiple-segments-level function.
+
+    Example usage:
+        train = {
+            "class": "PostprocessingDataset", "dataset": { "class": "..." },
+        }
+    """
+
+    def __init__(
+        self,
+        dataset: Dict[str, Any],
+        map_seq: Optional[Callable[[TensorDict], TensorDict]] = None,
+        map_seq_stream: Optional[Callable[[Iterator[TensorDict]], Iterator[TensorDict]]] = None,
```

Comment:
I think it might make sense that you also specify the expected output (to derive `num_outputs` from it, but also to make the automatic conversion easier, and also as an extra verification).

Maybe even the expected input?

(Similar as `extern_data` in the config.)